### PR TITLE
Added multiplication of transpose / adjoint matrices by diagonal matrices 

### PR DESCRIPTION
**Pull request by amontoison**
Updated version of #2482

### CODE REVIEW

**Comment by maleadt** [comment]
CI failures look related.

File: test/libraries/cublas.jl [hunk 1]
```diff
@@ -2293,6 +2293,30 @@ end
             d_Y = Diagonal(d_y)
             mul!(d_AY, d_A, d_Y)
             Array(d_AY) ≈ A * Diagonal(y)
+
+            YA = rand(elty,n,m)
+            d_YA = CuArray(XA)
```
Comment (amontoison) [inline test/libraries/cublas.jl]:
```suggestion
            d_YA = CuArray(YA)
```

**Comment by amontoison** [comment]
Julia wants to dispatch to another method with `Transpose` and `Adjoint`:
```julia
mul!(d_YA, d_Y, transpose(d_A))

If you want to allow scalar iteration, use `allowscalar` or `@allowscalar`
to enable scalar iteration globally or for the operations in question.
Stacktrace:
  [1] error(s::String)
    @ Base ./error.jl:35
  [2] errorscalar(op::String)
    @ GPUArraysCore ~/.julia/packages/GPUArraysCore/GMsgk/src/GPUArraysCore.jl:155
  [3] _assertscalar(op::String, behavior::GPUArraysCore.ScalarIndexing)
    @ GPUArraysCore ~/.julia/packages/GPUArraysCore/GMsgk/src/GPUArraysCore.jl:128
  [4] assertscalar(op::String)
    @ GPUArraysCore ~/.julia/packages/GPUArraysCore/GMsgk/src/GPUArraysCore.jl:116
  [5] getindex
    @ ~/.julia/packages/GPUArrays/qt4ax/src/host/indexing.jl:50 [inlined]
  [6] macro expansion
    @ ~/Applications/julia/julia-1.10.5/share/julia/stdlib/v1.10/LinearAlgebra/src/diagonal.jl:318 [inlined]
  [7] macro expansion
    @ ./simdloop.jl:77 [inlined]
  [8] __muldiag!(out::CuArray{Float64, 2, CUDA.DeviceMemory}, D::Diagonal{ComplexF64, CuArray{…}}, B::Transpose{ComplexF64, CuArray{…}}, _add::LinearAlgebra.MulAddMul{true, true, Bool, Bool})
    @ LinearAlgebra ~/Applications/julia/julia-1.10.5/share/julia/stdlib/v1.10/LinearAlgebra/src/diagonal.jl:317
  [9] _mul_diag!
    @ ~/Applications/julia/julia-1.10.5/share/julia/stdlib/v1.10/LinearAlgebra/src/diagonal.jl:391 [inlined]
 [10] _mul!
    @ ~/Applications/julia/julia-1.10.5/share/julia/stdlib/v1.10/LinearAlgebra/src/diagonal.jl:397 [inlined]
 [11] mul!
    @ ~/Applications/julia/julia-1.10.5/share/julia/stdlib/v1.10/LinearAlgebra/src/bidiag.jl:427 [inlined]
 [12] mul!(C::CuArray{Float64, 2, CUDA.DeviceMemory}, A::Diagonal{ComplexF64, CuArray{ComplexF64, 1, CUDA.DeviceMemory}}, B::Transpose{ComplexF64, CuArray{ComplexF64, 2, CUDA.DeviceMemory}})
    @ LinearAlgebra ~/Applications/julia/julia-1.10.5/share/julia/stdlib/v1.10/LinearAlgebra/src/matmul.jl:237
 [13] top-level scope
    @ REPL[71]:1
Some type information was truncated. Use `show(err)` to see complete types.
```
@dkarrasch Do you have an idea how I can fix it?
Should I add a method for `generic_matmatmul!` instead?

**Comment by dkarrasch** [comment]
Tests seem to pass? Or where is the problem? I think you're overloading the right method, which is the three-arg `mul!` in line [12]. That one is completely generic, so any method that specifies the arguments (to arrays that it owns) should have precedence.

**Comment by amontoison** [comment]
I forgot to change the branch when I tested the PR on a cluster...
The issue was between the chair and the screen (me)  :smile:
Thanks for confirming that I overloaded the right method.

**Pull request merged**